### ISSUE TITLE
Unable to set `readOnlyRootFilesystem` in Kubernetes deployment

### ISSUE THREAD

**Issue by onedr0p** [shlinkio/shlink#1967]
### Shlink version

3.7.2-non-root

### PHP version

3.7.2-non-root

### How do you serve Shlink

RoadRunner Docker image

### Database engine

PostgreSQL

### Database version

16.1

### Current behavior

I am trying to set a Kubernetes security context on shlink however it doesn't work and falis with the following error.

```yaml
            securityContext:
              allowPrivilegeEscalation: false
              readOnlyRootFilesystem: true
              capabilities: { drop: ["ALL"] }
```

```
│   ERR  In FileNotWritableException.php line 45:                                                                                                                                         │
│   ERR                                                                                                                                                                                   │
│   ERR    [ProxyManager\Exception\FileNotWritableException]                                                                                                                              │
│   ERR    Cannot rename "/tmp/ShlinkProxy__PM__GeoIp2DatabaseReaderGenerated1cc76100a                                                                                                    │
│   ERR    9790b1d2nlMain" to "/etc/shlink/data/proxies/ShlinkProxy__PM__GeoIp2Databas                                                                                                    │
│   ERR    eReaderGenerated1cc76100a9790b1d241b38c9b0c3801c.php": rename(/tmp/ShlinkPr                                                                                                    │
│   ERR    oxy__PM__GeoIp2DatabaseReaderGenerated1cc76100a9790b1d2nlMain,/etc/shlink/d                                                                                                    │
│   ERR    ata/proxies/ShlinkProxy__PM__GeoIp2DatabaseReaderGenerated1cc76100a9790b1d2                                                                                                    │
│   ERR    41b38c9b0c3801c.php): Read-only file system
```

### Expected behavior

The application should not need r/w access to the root filesystem

### How to reproduce

Deploy shlink in Kubernetes and set the security context as I mentioned above.

**Comment by acelaya** [comment]
Hmmm. Looks like some proxies are written to `tmp` first and then moved to Shlink's folder.

I'll check if this can be configured.

**Comment by onedr0p** [comment]
If they could stay in /tmp that would be great since I can mount an emptyDir to that location.

**Comment by acelaya** [comment]
> If they could stay in /tmp that would be great since I can mount an emptyDir to that location.

I'm afraid that's not possible (at least without an unreasonable amount of changes). Shlink needs to write a number of temporary/static files at runtime (GeoLite db file, entity proxies, service proxies, filesystem locks, configuration cache files, etc).

However, all files are (or should) be written inside `/etc/shlink/data/???`, so you should be able to mount `/etc/shlink/data` instead of `/tmp`, if that's an option.

Of course, that leaves the fact that there seems to be something writing in `/tmp` for some reason, which I need to address.

**Comment by onedr0p** [comment]
Good idea, so I created a emptyDir mount to `/etc/shlink/data` as well as `/tmp` however now I get this error:


```
│   ERR    Error Output:                                                                                                                                                                  │
│   ERR    ================                                                                                                                                                               │
│   ERR                                                                                                                                                                                   │
│   ERR    In InvalidDirectory.php line 15:                                                                                                                                               │
│   ERR                                                                                                                                                                                   │
│   ERR                                                                                                                                                                                   │
│   ERR      [Doctrine\Migrations\Finder\Exception\InvalidDirectory]                                                                                                                      │
│   ERR                                                                                                                                                                                   │
│   ERR      Cannot load migrations from "data/migrations" because it is not a valid d                                                                                                    │
│   ERR    ir                                                                                                                                                                             │
│   ERR      ectory
```

**Comment by acelaya** [comment]
The only directory you need to mount is `/etc/shlink/data`, but you need to make sure its contents are not replaced with an empty directory, because database migrations are located there as well.

I'm not sure if that's possible though. In retrospect, I think it probably would have made sense that migrations were located somewhere else, but I think the migrations tool tracks their path, so I'm not sure if that can be changed now. I'll check.

**Comment by acelaya** [comment]
> Hmmm. Looks like some proxies are written to tmp first and then moved to Shlink's folder.
>
> I'll check if this can be configured.

Writing to `/tmp` seems to be an implementation detail of a dependency down the line, from the DI container used by Shlink, which can generate those proxies. For some reason they write the file to `/tmp` and then try to move it to the configured destination.

Looking at the error, it seems writing to `/tmp` worked, and only moving the file to `/etc/shlink/data` failed, so I guess mounting the volume should cover this.

**Comment by onedr0p** [comment]
I tried mounting `/etc/shlink/data` to a [emptyDir](https://kubernetes.io/docs/concepts/storage/volumes/#emptydir) but had the error mentioned [here](https://github.com/shlinkio/shlink/issues/1967#issuecomment-1873035311)

**Comment by acelaya** [comment]
> I tried mounting `/etc/shlink/data` to a [emptyDir](https://kubernetes.io/docs/concepts/storage/volumes/#emptydir) but had the error mentioned [here](https://github.com/shlinkio/shlink/issues/1967#issuecomment-1873035311), as you mentioned it looks like it is clobbering that directory of all existing files.

I have already verified it is possible to move migrations somewhere else, so that everything inside the data dir is stuff generated at runtime.

I also need to check if it can be shipped as an empty dir, ensuring subdirs are created as required. At the moment, a few of those subdirs need to exist for everything to work, so mounting an empty dir would still fail even without taking migrations into consideration.

**Comment by acelaya** [comment]
I'm pretty confident that, as a workaround, you should be good to go by mounting empty dirs to `/etc/shlink/data/proxies` and `/etc/shlink/data/locks`.

If it still fails, you may need to mount `/tmp` as well.

In the meantime I'll continue investigating if it's possible to ship Shlink with an empty `/etc/shlink/data`.

**Comment by acelaya** [comment]
> I'm pretty confident that, as a workaround, you should be good to go by mounting empty dirs to `/etc/shlink/data/proxies` and `/etc/shlink/data/locks`.

Nah, scratch that. The GeoLite db downloading needs to write directly in `/etc/shlink/data` 🤦🏼

**Comment by onedr0p** [comment]
Looks like there is also a problem here... `The cache directory "data/cache" is not writable in /etc/shlink/vendor/mezzio/mezzio-fastroute/src/FastRouteRouter.php:558`

**Comment by acelaya** [comment]
> Looks like there is also a problem here... The cache directory "data/cache" is not writable in /etc/shlink/vendor/mezzio/mezzio-fastroute/src/FastRouteRouter.php:558

This is not needed when running Shlink with RoadRunner or openswoole. I'll change it as part of the other modifications.

**Comment by acelaya** [comment]
This is what I have finally decided to do.

1. Move migrations somewhere else. They are in fact source code, so the location was not correct anyway.
2. Update the docker entrypoint script to create the required `data` subdirs, only if they don't exist. This will allow to mount an empty dir at `/etc/shlink/data`. 
3. Keep everything else inside `data` dir as is. That way, this change won't affect others, specially those not using the docker image.

With this, the only consideration is the fact that one of the deps needs to write in `/tmp`, but I don't think that can be avoided.

Worst case scenario, you will have to mount `/tmp` as well, but according to the error above, my suspicion is that this is working already.

**Comment by acelaya** [comment]
I have just released v3.7.3-beta.1, which includes the changes explained above.

The `3.7.3-beta.1-non-root` docker image has been already published, if you want to try it.

**Comment by onedr0p** [comment]
Confirmed working with setting an emptyDir to `/tmp` and `/etc/shlink/data` 🎉
